python/beeai_framework/emitter/emitter.py

```diff
@@ -194,31 +194,40 @@ async def emit(self, name: str, value: Any) -> None:
             raise EmitterError.ensure(e)
 
     async def _invoke(self, data: Any, event: EventMeta) -> None:
-        executions: list[Coroutine[Any, Any, Any] | Task[Any]] = []
-        for listener in self._listeners:
-            if not listener.match(event):
-                continue
-
-            if listener.options and listener.options.once:
-                self._listeners.remove(listener)
-
-            async def run(ln: Listener = listener) -> Any:
-                try:
-                    if inspect.iscoroutinefunction(ln.callback):
-                        return await ln.callback(data, event)
-                    else:
-                        return ln.callback(data, event)
-                except Exception as e:
-                    raise EmitterError.ensure(
-                        e, message="One of the provided Emitter callbacks has failed.", event=event
-                    )
+        tasks: list[Task[Any]] = []
+
+        async def run(ln: Listener) -> Any:
+            try:
+                ln_async = ensure_async(ln.callback)
```

Comment:
callback here must be async or return awaitable right? Or it can be a sync function?